src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -779,9 +779,10 @@ private static void ReportUnaryOperatorError(CSharpSyntaxNode node, BindingDiagn
             Error(diagnostics, errorCode, node, operatorName, operand.Display);
         }
 
-        private void ReportAssignmentOperatorError(AssignmentExpressionSyntax node, BindingDiagnosticBag diagnostics, BoundExpression left, BoundExpression right, LookupResultKind resultKind)
+        private void ReportAssignmentOperatorError(AssignmentExpressionSyntax node, BinaryOperatorKind kind, BindingDiagnosticBag diagnostics, BoundExpression left, BoundExpression right, LookupResultKind resultKind)
         {
-            if (((SyntaxKind)node.OperatorToken.RawKind == SyntaxKind.PlusEqualsToken || (SyntaxKind)node.OperatorToken.RawKind == SyntaxKind.MinusEqualsToken) &&
+            if (IsTypelessExpressionAllowedInBinaryOperator(kind, left, right) &&
```

Comment:
Looks like `IsTypelessExpressionAllowedInBinaryOperator` returns `false` if the expression is a kind of typeless expression which is disallowed in binary operator and `true` if it's allowed or not typeless at all, right? The name confused me as it sounds like it would return `true` only if the expression is typeless. Should it perhaps be renamed to
- `IsDisallowedTypelessExpressionInBinaryOperator` and its semantics negated, or
- `IsExpressionAllowedInBinaryOperator` (with semantics unchanged)?

Also, what happened when this check wasn't here? The `GenerateImplicitConversionError` failed?

Reply:
I think the current name is appropriate, looking at its usage in `BinaryOperatorOverloadResolution`:
```
        private BinaryOperatorAnalysisResult BinaryOperatorOverloadResolution(...)
        {
            if (!IsTypelessExpressionAllowedInBinaryOperator(kind, left, right))
            {
                resultKind = LookupResultKind.OverloadResolutionFailure;
                originalUserDefinedOperators = default;
                return default(BinaryOperatorAnalysisResult);
            }
            ...
```

> Also, what happened when this check wasn't here? 

In DEBUG, the program would assert below (`Debug.Assert(!conversion.IsImplicit);`).  
In RETAIL, the program would fail to report an error and would crash during emit phase. I didn't investigate in details, but I think it's just that `GenerateImplicitConversionError` doesn't handle the case and so didn't report an error.

Reply:
> looking at its usage in `BinaryOperatorOverloadResolution`

Yeah, I guess there it makes more sense since it's negated.